### PR TITLE
qemu_guest_agent:Add check step after thaw FS in guest

### PR DESCRIPTION
**Pull request by xiagao**
1.add check step after thaw fs in guest.
2.fix bug about 'thread._local' object has no attribute 'contexts'

id: 1543259 , 1508356
Signed-off-by: Xiaoling Gao xiagao@redhat.com

### CODE REVIEW

**Comment by xiagao** [comment]
'1-smp_8.256m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_fsfreeze' Finshed 2018-03-07 04:17:13 2018-03-07 04:19:32 PASS 139.081751

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -802,29 +802,39 @@ def _action_before_fsfreeze(self, *args):
...
-            raise error.TestError("Could not find any opened session")
+            self.test.error("Could not find any opened session")
         # Use the last opened session to send cmd.
         session = self._open_session_list[-1]
+
         try:
             session.cmd(self.params["gagent_fs_test_cmd"])
         except aexpect.ShellTimeoutError:
-            logging.debug("FS freeze successfully.")
+            logging.info("FS is frozen,can't write in guest.")
```
Comment (kimi1978wy) [inline qemu/tests/qemu_guest_agent.py]: batter add "it is expected" in logging.info, so will not misunderstand when checking log.

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -871,9 +882,6 @@ def gagent_check_fsfreeze(self, test, params, env):
                              " detail: '%s'", detail)
             raise
```
Comment (kimi1978wy) [inline qemu/tests/qemu_guest_agent.py]: Can we omit this "raise"?

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: Better to leave it here, raise exception and stop executing the following test.

File: qemu/tests/qemu_guest_agent.py (hunk 1 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: thanks for your advice, update.
@kimi1978wy  help to check it again,thx.

**Comment by xiagao** [comment]
@hereischen  could you help review it? and I also deprecate Auotest modules.

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -836,31 +841,32 @@ def gagent_check_fsfreeze(self, test, params, env):
...
         try:
             expect_status = self.gagent.FSFREEZE_STATUS_THAWED
             self.gagent.verify_fsfreeze_status(expect_status)
         except guest_agent.VAgentFreezeStatusError:
             # Thaw guest FS if the fs status is incorrect.
             self.gagent.fsthaw(check_status=False)
 
-        self._action_before_fsfreeze(test, params, env)
+        self._action_before_fsfreeze()
+        error_context.context("Freeze the FS.")
```
Comment (YanhuiMa) [inline qemu/tests/qemu_guest_agent.py]: lack of 'logging.info'.

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: it's ok without logging.info like line867. 
05:12:12 INFO | Context: Freeze the FS. --> Verfiy FS is frozen in guest.

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -792,8 +787,8 @@ def find_interface_by_name(interface_list, target_interface):
                        interfaces_pre_add]
         if (len(bridge_list) != 1) or \
            ("lo_brige" != bridge_list[0]["name"]):
-            error.TestFail("the interface list info after interface was down"
-                           " was not as expected")
+            test.fail("the interface list info after interface was down"
+                      "was not as expected")
```
Comment (YanhuiMa) [inline qemu/tests/qemu_guest_agent.py]: should not remove blank.

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -206,8 +205,8 @@ def gagent_verify(self, params, vm):
         error_context.context("Check if guest agent work.", logging.info)
 
         if not self.gagent:
-            raise error.TestError("Could not find guest agent object"
-                                  "for VM '%s'" % vm.name)
+            self.test.error("Could not find guest agent object "
+                            "for VM '%s'" % vm.name)
```
Comment (YanhuiMa) [inline qemu/tests/qemu_guest_agent.py]: the blank

File: qemu/tests/qemu_guest_agent.py [hunk 6]
```diff
@@ -802,29 +797,39 @@ def _action_before_fsfreeze(self, *args):
...
+        error_context.context("Verify FS is thawed in guest.", logging.info)
+
+        if not self._open_session_list:
+            self.test.error("Could not find any opened session.")
+        # Use the last opened session to send cmd.
+        session = self._open_session_list[-1]
+
+        try:
+            session.cmd(self.params["gagent_fs_test_cmd"])
+        except aexpect.ShellTimeoutError:
```
Comment (hereischen) [inline qemu/tests/qemu_guest_agent.py]: It seems the lines 820 ~ 829 are same with lines in `_action_after_fsfreeze`.

**Comment by hereischen** [comment]
Acked-by: Haotong Chen hachen@redhat.com

File: qemu/tests/qemu_guest_agent.py (hunk 6 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: To make the four functions about fsfreeze more clear, I think it's ok. Thanks.

File: qemu/tests/qemu_guest_agent.py (hunk 4 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: add a blank space,thx

**Comment by xiagao** [comment]
@luckyh could you pls have a check?

File: qemu/tests/qemu_guest_agent.py [hunk 7]
```diff
@@ -802,29 +797,39 @@ def _action_before_fsfreeze(self, *args):
...
     def _action_before_fsthaw(self, *args):
         pass
 
     @error_context.context_aware
     def _action_after_fsthaw(self, *args):
-        pass
+        error_context.context("Verify FS is thawed in guest.", logging.info)
+
+        if not self._open_session_list:
+            self.test.error("Could not find any opened session.")
```
Comment (luckyh) [inline qemu/tests/qemu_guest_agent.py]: I think this shouldn't be an error, is there some reason to do so?

File: qemu/tests/qemu_guest_agent.py [hunk 8]
```diff
@@ -836,31 +841,32 @@ def gagent_check_fsfreeze(self, test, params, env):
...
+                              logging.info)
         try:
             expect_status = self.gagent.FSFREEZE_STATUS_THAWED
             self.gagent.verify_fsfreeze_status(expect_status)
         except guest_agent.VAgentFreezeStatusError:
             # Thaw guest FS if the fs status is incorrect.
             self.gagent.fsthaw(check_status=False)
 
-        self._action_before_fsfreeze(test, params, env)
+        self._action_before_fsfreeze()
```
Comment (luckyh) [inline qemu/tests/qemu_guest_agent.py]: Should keep these arguments, also apply for below ones.

File: qemu/tests/qemu_guest_agent.py (hunk 7 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: I modify it with creating a session.

File: qemu/tests/qemu_guest_agent.py [hunk 9]
```diff
@@ -937,7 +939,7 @@ def get_qga_pkg_path(self, test, session, params, vm):
 
             error_context.context("Download qemu-ga.msi from website and copy "
                                   "it to guest.", logging.info)
-            utils.run(gagent_download_cmd, float(params.get("login_timeout", 360)))
+            process.system_output(gagent_download_cmd, float(params.get("login_timeout", 360)))
```
[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: modified

**Comment by xiagao** [comment]
@luckyh updated, pls have a check, thank you.

**Comment by xiagao** [comment]
@luckyh  updated, thx
(1/1) smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_fsfreeze: PASS (115.39 s)

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**